### ISSUE TITLE
Switch to webpack

### ISSUE THREAD

**Issue by j-f1** [skulpt/skulpt#666]
[webpack](https://webpack.js.org) is a new module builder for JavaScript. 

[Comparison with other module builders](https://webpack.js.org/guides/why-webpack/#comparison)
If we decide to switch, here is a checklist:

- [ ] add webpack to the list of dependencies
- [ ] replace all of the `goog.*` functions with native JS
- [ ] replace `Sk.foo.bar.baz` with ES6 imports (webpack will transpile these automatically)
- [ ] make a backwards-compatibility module that creates an `Sk` object with all of the appropriate attributes

<sup>I’m happy to do this work myself, I just don’t want to waste effort creating something nobody here wants.</sup>

**Comment by waywaaard** [comment]
First of all I like the idea. But there are several problems that come to my mind:

- The skulpt code is pretty tight coupled with google closure, e.g. see `/gen/*` or make a search for all `goog.*` occurences
- Our inheritance code relies currently on the google closure behavior, though that should be not a problem to switch out
- Our checking also relies on that
- Currently, we have order problems, so sometimes we rely on the existance of functions at later points
- And I am not sure if we introduce a performance decrease with webpack. Closure compiler is pretty good, and the comparison table lacks google closure

I guess for a project of this size, typescript would be more suited. Yet, a proper review and testing process (which is sometimes quite difficult in skulpt [internals]) catches most of those.

**Comment by j-f1** [comment]
* We can switch to Node/npm for most of the `goog.*` occurrences, and `src/str.js` could use its own regexes.
  <details><summary>Details</summary><ul>
  <li>`goog.isArrayLike` → `isArrayLike` [from Lodash](https://lodash.com/docs#isArrayLike) [npm](https://www.npmjs.com/package/lodash.isarraylike)
  <li>`goog.global` → `global` (set to `window` by webpack automatically)
  <li>`goog.string.*` → regexes (they’re only used in `src/str.js`)
  <li>`goog.string.trimLeft` → [`trim-left`](https://www.npmjs.com/package/trim-left)
  </ul></details>
* Inheritance could be refactored with ES6 classes.
* All assert logic can be replaced with the Node standard library
  <details><summary>Details</summary><ul>
  <li> add `import assert from 'assert'` to the top of all files that use `assert`
  <li>`goog.asserts.*` → `assert.*`
  <li>`goog.asserts.assert(...)` → `assert(...)`
  <li>`goog.asserts.AssertionError` isn’t used anywhere besides `gen/closure_ctor_hack.js`
  </ul></details>
* If you need access to a function somewhere else, you could just `export` it, then `import` it where it’s needed.
* webpack uses UglifyJS, which has [a comparison](http://lisperator.net/uglifyjs/#stats) with Google Closure. An advantage of webpack/Uglify is that they don’t require Java because they’re pure JS.

webpack supports asynchronous loading so the Skulpt stdlib could be separated into several chunks to reduce loading time, and it supports [*tree shaking*](https://webpack.js.org/guides/tree-shaking/) which could, if Skulpt was built into a webpack project, allow for unused methods to be eliminated at the per-export level.

**Comment by waywaaard** [comment]
- Okay, but this will be huge amount of work to do.
- I am still not sure, if we can catch every edgecase with ES6 Classes
- Yeah, I already though about the async loading of the stlib modules.
- I guess the export/import would make the dependencies in each class more visible
- Tree Shaking is still broken in webpack v2, and we only need this for external libs.
- The differences between uglifyjs and closure are very marginal in terms of saved bytes. However, it seems to be faster

**Comment by albertjan** [comment]
Definitely love this idea!, But I think we're skipping a very important part of the google closure compiler here, type-checking. Before we rush down one route I would like to investigate some other too. Personally I haven't used web pack but I have used rollup. And these tools are very powerful and would help us greatly in the problems we have with load order. 

With regard to the type-checking there are two solutions out there at the moment (that I know of). 

- typescript 
- flow

I haven't worked with any of these at length. So I can't say which is better. Can these be combined with tools like webpack and rollup.

About the tree shaking, we do a lot of string based lookups in skulpt tree-shaking might shake off too much. 😄

**Comment by j-f1** [comment]
@waywaaard I was thinking that a Skulpt loader could be created for webpack, which would automatically compile Python so that tree shaking could remove what isn’t needed in the compiled file.

@albertjan TypeScript can be enabled with [typescript-loader](https://www.npmjs.com/package/typescript-loader), and a cursory search shows [flowtype-loader](https://www.npmjs.com/package/flowtype-loader) for Flow. Also, webpack is pretty smart about evaluating expressions in `require`, so it includes everything necessary.

**Comment by albertjan** [comment]
I don't know how @bnmnetp and @rixner feel about this. I'm sure you'll have prove we don't loose any performance with this. I do think type-checking will be an important part of this change.

**Comment by albertjan** [comment]
![img](https://cloud.githubusercontent.com/assets/883126/12903661/453405a2-ce7e-11e5-9ddf-389f4ef8f60f.png) we will likely suffer a size hit when we move away from closure.

**Comment by albertjan** [comment]
We should investigate rollup too in this example it performs better in size and speed. (if slightly). It's probably not that difficult to switch packers.

**Comment by albertjan** [comment]
https://github.com/samccone/The-cost-of-transpiling-es2015-in-2016 <-- this is where I got the info from.

**Comment by bnmnetp** [comment]
It's not clear to me how this really helps us move skulpt forward?  Maybe someone can help me see the light.

 We are all going to spend our time and energy working on the things we are excited about, but for me, putting resource into implementing more of the stdlib or more work on internals that enable us to make progress on just including more pure python elements of the CPython stdlib would have much larger benefit.

**Comment by rixner** [comment]
I concur with @bnmnetp .  This feels like a "if it ain't broke, don't fix it" kind of situation.  I am concerned about diverting time and energy into this when I haven't seen any compelling evidence that it makes anything better.

As with @bnmnetp , I believe that everyone's time could be better spent adding capabilities, or whatever other elements of the internals and stdlib that gets people motivated,  or even moving towards Python3.

I am concerned with (1) the possible slowdowns this might cause, (2) the bloat in size of the resulting distribution, and (3) the cognitive load on myself and all of the other developers to switch to a different way of doing things.  Closure may not be the greatest system, but it does do some type checking, very effective minimization, and some performance optimization.

**Comment by j-f1** [comment]
That’s fine with me — I’m not the one who has invested lots of effort into creating Skulpt 😄

**Comment by albertjan** [comment]
@j-f1 I don't know if you should have closed this issue so fast. There are definite advantages to moving away from closure-compiler. I think they should be listed.

**Comment by bnmnetp** [comment]
Thanks @albertjan -- I did say I was willing to be enlightened. 😄  (I know my Runestone project is in dire need of something like this!). If this is a win for skulpt, and it brings new contributors to the project then its all good.

**Comment by rixner** [comment]
@j-f1, I also said that "I haven't seen any compelling evidence that it makes anything better".  That doesn't mean there isn't any...

**Comment by albertjan** [comment]
Pro:
- Use es6 syntax, arrow functions let and const variables, es6 classes.
- module loading, no more rigid loading of files. webpack (or other tool) will take care of that
- no more v8 dependency we can run on node. (I've been testing this and it does run on node already, just not very elegantly)
- quicker build process.
- If we use typescript, a much finer grained optional type system and an editor (vscode that will really help you) 
- Likely as fast or even a bit faster
- It's "the future"

Cons:
- We'll all have to learn es6, webpack and typescript
- Likely a bit bigger
- A serious amount of work (we'd need @j-f1 commitment to the project)

I'll keep these lists updated. As more points come in.

**Comment by j-f1** [comment]
@albertjan Another pro: in the future, we could potentially make a `skulpt-loader` for webpack that would transpile Python into JS, allowing people to publish Python libraries to npm, and also allow the use of Python in a website’s codebase wherever convenient.

**Comment by albertjan** [comment]
@j-f1 Added, we'd need to do something like PR #609. There are certain runtime settings that do affect compilation, so this won't work for everyone.

**Comment by bnmnetp** [comment]
@albertjan Does the closure library or something else in Skulpt keep us from using the new ES6 stuff right now?  Lots to like in ES6 as I prepare to teach Javascript over the next few weeks.

**Comment by albertjan** [comment]
Yes, because you need to transpile es6(or es2015) to es5. We'd need the transpiler to output code in a way closure understands. Which will be difficult because these systems generally create all sorts of functions on the fly to deal with the translation, and these will not have the necessary annotations for closure.

**Comment by albertjan** [comment]
@j-f1 I actually tried this... It's near impossible 😄 This is the dependency graph according to Webpack. I'll open a PR with what I have now 😁 
![graph](https://user-images.githubusercontent.com/301791/56787173-a4234700-67f3-11e9-955d-1d63c74f933a.png) See PR #857
